pareto/strategic_water_management/strategic_produced_water_optimization.py

```diff
@@ -6170,174 +6247,1550 @@ def ObjectiveFunctionRule(b):
+
+    # region Completions Pad
+
+    # Water that is Piped and Trucked to a completions pad is mixed and split into two output streams.
+    # Stream (1) goes to the completions pad and stream (2) is input to the completions storage.
+    # This is the intermediate step.
+    # Finally, water that meets completions demand comes from two inputs.
+    # The first input is output stream (1) from the intermediate step.
+    # The second is outgoing flow from the storage tank.
+
+    def CompletionsPadIntermediateWaterQuality(b, p, w, t):
+        return sum(
+            sum(
+                model.v_F_DiscretePiped[n, p, t, w, q] * model.p_discrete_quality[w, q]
+                for q in model.s_Q
+            )
+            for n in model.s_N
+            if model.p_NCA[n, p]
+        ) + sum(
+            model.v_F_Piped[p_tilde, p, t] * b.p_nu_pad[p, w]
+            for p_tilde in model.s_PP
+            if model.p_PCA[p_tilde, p]
+        ) + sum(
+            sum(
+                model.v_F_DiscretePiped[s, p, t, w, q] * model.p_discrete_quality[w, q]
+                for q in model.s_Q
+            )
+            for s in model.s_S
+            if model.p_SCA[s, p]
+        ) + sum(
+            model.v_F_Piped[p_tilde, p, t] * b.p_nu_pad[p, w]
+            for p_tilde in model.s_CP
+            if model.p_CCA[p_tilde, p]
+        ) + sum(
+            sum(
+                model.v_F_DiscretePiped[r, p, t, w, q] * model.p_discrete_quality[w, q]
+                for q in model.s_Q
+            )
+            for r in model.s_R
+            if model.p_RCA[r, p]
+        ) + sum(
+            model.v_F_Sourced[f, p, t] * b.p_nu_freshwater[f, w]
+            for f in model.s_F
+            if model.p_FCA[f, p]
+        ) + sum(
+            model.v_F_Trucked[p_tilde, p, t] * b.p_nu_pad[p, w]
+            for p_tilde in model.s_PP
+            if model.p_PCT[p_tilde, p]
+        ) + sum(
+            model.v_F_Trucked[p_tilde, p, t] * b.p_nu_pad[p, w]
+            for p_tilde in model.s_CP
+            if model.p_CCT[p_tilde, p]
+        ) + sum(
+            sum(
+                model.v_F_DiscreteTrucked[s, p, t, w, q]
+                * model.p_discrete_quality[w, q]
+                for q in model.s_Q
+            )
+            for s in model.s_S
+            if model.p_SCT[s, p]
+        ) + sum(
+            model.v_F_Trucked[f, p, t] * b.p_nu_freshwater[f, w]
+            for f in model.s_F
+            if model.p_FCT[f, p]
+        ) <= sum(
+            model.v_F_DiscreteFlowCPIntermediate[p, t, w, q]
+            * model.p_discrete_quality[w, q]
+            for q in model.s_Q
+        )
+
+    model.CompletionsPadIntermediateWaterQuality = Constraint(
+        model.s_CP,
+        model.s_W,
+        model.s_T,
+        rule=CompletionsPadIntermediateWaterQuality,
+        doc="Completions pad water quality",
+    )
+
+    # The flow to the completion pad is given, so the quality can be continuous.
+    model.v_Q_CompletionPad = Var(
```

Comment:
On github, I still see no units for this variable.

Reply:
Should be there now